File: src/speech.cpp
#include <frasier_utilities/speech.h>

Speech::Speech(ros::NodeHandle n, bool debug) :
	nh_(n),
	speech_cli_(SPEECH_CLIENT_TOPIC, true),
	debug_(debug)
{
	bool speech_cli_running = speech_cli_.waitForServer(ros::Duration(3.0));
	if (speech_cli_running)
		std::cout << "SPEECH CLIENT: Speech initialized." << std::endl;
	else
		std::cout << "SPEECH CLIENT: Speech is NOT initialized!" << std::endl;

	// Voice setup
	voice_msg_.language = tmc_msgs::Voice::kEnglish;
	voice_msg_.queueing = true;
	voice_msg_.interrupting = true;
	voice_msg_.sentence = "Start";
	voice_goal_.data = voice_msg_;
}

void Speech::sendCurrentGoal(bool wait) {
	speech_cli_.sendGoal(voice_goal_);
	if (wait)
		speech_cli_.waitForResult();
}

void Speech::setSentence(std::string sentence) {
	voice_msg_.sentence = sentence;
	voice_goal_.data = voice_msg_;
}

void Speech::say(std::string sentence, bool wait) {
	setSentence(sentence);
	sendCurrentGoal(wait);
}

File: src/frasier_utilities/speech.py
#!/usr/bin/env python
# ROS
import rospy
import actionlib

# HSR
from tmc_msgs.msg import Voice, TalkRequestAction, TalkRequestGoal
from hsr_constants import *


class Speech(object):
    def __init__(self):
        # Voice Msg
        self.voice_pub = rospy.Publisher(SPEECH_TOPIC, Voice, queue_size=10)
        self._speech_cli = actionlib.SimpleActionClient(SPEECH_CLIENT_TOPIC, TalkRequestAction)
        rospy.loginfo("SPEECH CLIENT: Waiting for speech action server...")
        speech_is_running = self._speech_cli.wait_for_server(rospy.Duration(2))

        if speech_is_running:
            rospy.loginfo("SPEECH CLIENT: Speech initialized.")
        else:
            rospy.loginfo("SPEECH CLIENT: Speech is NOT initialized!")
        self.voice_msg = Voice()
        self.voice_msg.language = Voice.kEnglish
        self.voice_msg.queueing = True
        self.voice_msg.interrupting = True
        self.voice_msg.sentence = 'Start'
        self.goal = TalkRequestGoal()

    def send_current_goal(self, blocking=True):
        self._speech_cli.send_goal(self.goal)
        if blocking:
            self._speech_cli.wait_for_result()

    def set_sentence(self, text):
        self.voice_msg.sentence = text
        self.goal.data = self.voice_msg

    def say(self, text, blocking=True):
        self.set_sentence(text)
        self.send_current_goal(blocking)
